### PR TITLE
Add provisional support for real devices

### PR DESCRIPTION
**Pull request by imurchie**
Basic support for real devices. NO functionality was added beyond allowing a session to be created.

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -185,37 +185,43 @@ class XCUITestDriver extends BaseDriver {
...
-        sim = await this.createSim();
-        udid = sim.udid;
+        device = await this.createSim();
+        udid = device.udid;
         realDevice = false;
       }
     }
 
-    return { sim, udid, realDevice };
+    return { device, udid, realDevice };
```
Comment (jlipps) [inline lib/driver.js]: just for readability it could be nice to flatten this section out:

```js
if (this.opts.udid && await simExists(this.opts.udid)) {
  ...
  return {device, udid, realDevice};
}

if (this.opts.udid) {
  ...
  return {device, udid, realDevice};
}

device = await getExistingSim(this.opts.deviceName, this.opts.platformVersion);

if (device) {
  ...
  return {device, udid, realDevice};
}

log.info(`udid not provided...`);
...
return {device, udid, realDevice};
```

It's up to you; there's a little more repetition in the `return`s, but I find multiple embedded conditionals hard to follow especially on small screen sizes. you have to keep more state in your head.

Comment (imurchie) [inline lib/driver.js]: Yeah, I'm all for clarity over strict DRY-ness. I have a few other refactory stuff to do, and will make this change.

File: lib/utils.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  } catch (err) {
+    log.debug('Using udidetect');
+    cmd = require.resolve('udidetect');
+  }
+  let udid;
+  try {
+    let {stdout} = await exec(cmd, args, {timeout: 3000});
+    udid = stdout.split('\n')[0];
+  } catch (err) {
+    log.error(`Error detecting udid: ${err.message}`);
```
Comment (jlipps) [inline lib/utils.js]: log.errorAndThrow?

File: lib/webdriveragent.js [hunk 3]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
+          await systemLogExists(this.device);
         }
-      });
 
-      log.info(`Waiting for WebDriverAgent to start on device`);
-      await agentStartedOnDevice;
-      log.info(`WebDriverAgent started at url ${agentUrl}`);
+        let agentUrl;
+        let lineCount = 0;
+        let reachedEnd = realDevice ? false : true; // simulator does not need to wait
```
Comment (jlipps) [inline lib/webdriveragent.js]: `!!realDevice`

File: lib/webdriveragent.js [hunk 4]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
+        let agentStartedOnDevice = this.deviceLogs.start((stdout) => {
```
Comment (jlipps) [inline lib/webdriveragent.js]: rather than awaiting the bare promise below, this might look a little better like:

```js
const startDetector = (stdout) => {
  ...
}

await this.deviceLogs.start(startDetector);
```

File: lib/webdriveragent.js [hunk 5]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
+    this.device = args.device;
     this.platformVersion = args.platformVersion;
     this.host = args.host;
     this.agentPath = args.agentPath || AGENT_PATH;
     this.bootstrapPath = args.bootstrapPath || BOOTSTRAP_PATH;
   }
 
-  async launch (sessionId) {
+  async launch (sessionId, realDevice = false) {
     return new B(async (resolve, reject) => {
```
Comment (jlipps) [inline lib/webdriveragent.js]: why is this all wrapped in a promise constructor?

File: lib/webdriveragent.js [hunk 6]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
+  }
+
+  setupXcodeLogging (xcodebuild, reject) {
+    xcodebuild.on('output', (d, e) => {
+      // we want to pull out the log file that is created, and highlight it
+      // for diagnostic purposes
+      if (e.indexOf('Writing diagnostic log for test session to') !== -1) {
+        // pull out the first line that begins with the path separator
+        // which should be the line indicating the log file generated
+        this.logLocation = _.first(_.remove(e.trim().split('\n'), (v) => v.indexOf(path.sep) === 0));
```
Comment (jlipps) [inline lib/webdriveragent.js]: i've been preferring the model of managing state at as high a level as possible in the application. on that model, we'd return `logLocation` and if it needs to be managed as state, that would happen in the calling method, or that method's calling method, as high as possible. in many cases i've discovered we don't even need to store something on `this`.

File: lib/webdriveragent.js [hunk 7]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
 
   async checkForDependencies () {
-    if (!await fs.hasAccess(this.bootstrapPath + "/Carthage")) {
+    if (!await fs.hasAccess(`${this.bootstrapPath}/Carthage`)) {
       await exec('/bin/bash', ['Scripts/bootstrap.sh'], {cwd: this.bootstrapPath});
     }
   }
 
-  createXcodeBuildSubProcess () {
+  createXcodeBuildSubProcess (reject, realDevice = false) {
```
Comment (jlipps) [inline lib/webdriveragent.js]: I have to say I'm not a big fan of this pattern of passing `reject` around, and feel like there must be a more transparent way...

File: lib/webdriveragent.js [hunk 8]
```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
...
+    return logs;
+  }
+
+  setupLogging (logs, prefix) {
+    function isPertinentLogLine (line) {
+      return line.length &&
+            (line.indexOf(AGENT_LOG_PREFIX) !== -1 ||
+             line.indexOf(AGENT_RUNNER_LOG_PREFIX) !== -1);
+    }
+    logs.on('output', (d, e) => {
```
Comment (jlipps) [inline lib/webdriveragent.js]: are `d` and `e` stdout and stderr? if so, (a) these variables are not necessarily that transparent, and (b) i'm pretty sure you never get both at the same time, so you might be able to compress this:

```js
let line = d || e;
if (isPertinentLogLine(line)) {
  agentLog.debug(`${prefix}: ${line.trim()}`);
}
```

**Comment by jlipps** [comment]
wow, this was a lot! in general 👍  for the change, have some comments on how it's put together though.

File: lib/webdriveragent.js (hunk 5 above)
Comment (imurchie) [inline lib/webdriveragent.js]: To be honest, I don't really know. I've just left the shell of this as Jonah made it. But if it weren't it would clean up error handling, etc, too (like the `reject` being passed around). I'll try shifting things out and see if it continues to work.

Comment (jlipps) [inline lib/webdriveragent.js]: I think it's worth cleaning up as part of this PR, even though it was pre-existing

**Comment by Shanayadg** [comment]
New to the group -
I updated appium to beat 1.6.0-beta@3 and UI version of appium is still 1.5.3.
able to install ipa file created by xcode 8 on ios 10 device , but unable to launch  and below is the error  , please help me on this 
MJSONWP] Encountered internal error running command: Error: Could not find Automation.tracetemplate in any of the followinglocations /Applications/Xcode-beta.app/Contents/Applications/Instruments.app/Contents/PlugIns/AutomationInstrument.xrplugin/Contents/Resources/Automation.tracetemplate,/Applications/Xcode-beta.app/Contents/Applications/Instruments.app/Contents/PlugIns/AutomationInstrument.bundle/Contents/Resources/Automation.tracetemplate

**Comment by imurchie** [comment]
You're not running Appium 1.6.0.
